### DISCUSSION TITLE
Passing a callback to JS function in Static Hermes

### DISCUSSION THREAD

**Discussion by luboslenco** [facebook/hermes#1269] [Static Hermes]
Hello,

Is there a way to pass a JS function to C function? Afterwards I would like to call this JS function from C.

```js
let set_callback = $SHBuiltin.extern_c({}, function _set_callback(callback: c_ptr): void {});

function my_callback() {
}

set_callback(my_callback);
```

Right now this will result in `error: ft: function parameter 'callback' type mismatch`.

**Comment by tmikov** [comment]
Hi @luboslenco, that's a great question. Support for JS callbacks is planned (we have some unpublished code for it), but is not currently possible directly (I believe it can still be emulated using JSI).

To execute, every JS function needs two pieces of context:
1. A pointer to the JS runtime.
2. A pointer to its lexical environment (part of the "closure").

These are automatically passed around in JS, but when calling from native, they need to come from somewhere.

In JSI parlance, the former is `jsi::Runtime` and the latter is `jsi::Function`. You need access to both at the point of invokation the callback.

These are the steps to achieve this:
- The callback signature needs to take a context parameter.
- The context points to a heap allocated structure, storing the closure and the runtime.
- There is a wrapper with the callback signature, that invokes the JS function.
